modules/tests/test_opendaq_device_modules/test_opcua_device_modules.cpp

```diff
@@ -484,6 +492,69 @@ TEST_F(OpcuaDeviceModulesTest, SdkPackageVersion1)
     ASSERT_EQ(info.getPropertyValue("sdkVersion"), OPENDAQ_PACKAGE_VERSION);
 }
 
+TEST_F(OpcuaDeviceModulesTest, AuthenticationDefault)
+{
+    auto serverInstance = CreateServerInstance();
+    auto clientInstance = InstanceBuilder().build();
+
+    auto config = clientInstance.getAvailableDeviceTypes().get("opendaq_opcua_config").createDefaultConfig();
+    config.setPropertyValue("Username", "");
+    config.setPropertyValue("Password", "");
+
+    auto device = clientInstance.addDevice("daq.opcua://127.0.0.1", config);
+    ASSERT_TRUE(device.assigned());
+}
+
+TEST_F(OpcuaDeviceModulesTest, AuthenticationDefinedUsers)
+{
+    auto users = List<IUser>();
+    users.pushBack(User("jure", "jure123"));
+    users.pushBack(User("tomaz", "tomaz123"));
+
+    auto authenticationProvider = StaticAuthenticationProvider(false, users);
+    auto serverInstance = CreateServerInstance(authenticationProvider);
+
+    auto clientInstance = InstanceBuilder().build();
+    auto config = clientInstance.getAvailableDeviceTypes().get("opendaq_opcua_config").createDefaultConfig();
+
+    ASSERT_THROW(clientInstance.addDevice("daq.opcua://127.0.0.1", config), DaqException);
+
+    config.setPropertyValue("Username", "jure");
+    config.setPropertyValue("Password", "wrongPass");
+    ASSERT_THROW(clientInstance.addDevice("daq.opcua://127.0.0.1", config), DaqException);
+
+    config.setPropertyValue("Username", "andrej");
+    config.setPropertyValue("Password", "andrej123");
+    ASSERT_THROW(clientInstance.addDevice("daq.opcua://127.0.0.1", config), DaqException);
+
+    config.setPropertyValue("Username", "jure");
+    config.setPropertyValue("Password", "jure123");
+    auto device = clientInstance.addDevice("daq.opcua://127.0.0.1", config);
+    ASSERT_TRUE(device.assigned());
+    clientInstance.removeDevice(device);
+
+    config.setPropertyValue("Username", "tomaz");
+    config.setPropertyValue("Password", "tomaz123");
+    device = clientInstance.addDevice("daq.opcua://127.0.0.1", config);
+    ASSERT_TRUE(device.assigned());
+    clientInstance.removeDevice(device);
+}
+
+TEST_F(OpcuaDeviceModulesTest, AuthenticationAllowNoOne)
+{
+    auto authenticationProvider = AuthenticationProvider(false);
+    auto serverInstance = CreateServerInstance(authenticationProvider);
+
+    auto clientInstance = InstanceBuilder().build();
+    auto config = clientInstance.getAvailableDeviceTypes().get("opendaq_opcua_config").createDefaultConfig();
+
+    ASSERT_THROW(clientInstance.addDevice("daq.opcua://127.0.0.1", config), DaqException);
+
+    config.setPropertyValue("Username", "jure");
+    config.setPropertyValue("Password", "jure123");
+    ASSERT_THROW(clientInstance.addDevice("daq.opcua://127.0.0.1", config), DaqException);
```

Comment:
I am planning to rework opcua client, so AuthenticationFailed exception can be thrown here.